Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.mm

```diff
@@ -485,17 +485,25 @@ static inline void appendEndsPreviousSampleDurationMarker(AVAssetWriterInput *as
         return;
     }
 
-    flushCompressedSampleBuffers([weakThis = WeakPtr { *this }]() mutable {
-        if (!weakThis)
-            return;
+    if (m_videoCompressor)
+        m_videoCompressor->flush();
+    if (m_audioCompressor)
+        m_audioCompressor->flush();
+
+    // We hop to the main thread since flushing the video compressor might trigger starting the writer asynchronously.
+    callOnMainThread([this, weakThis = WeakPtr { *this }]() mutable {
+        flushCompressedSampleBuffers([weakThis = WTFMove(weakThis)]() mutable {
+            if (!weakThis)
+                return;
 
-        ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-        [weakThis->m_writer flush];
-        ALLOW_DEPRECATED_DECLARATIONS_END
+            ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+            [weakThis->m_writer flush];
+            ALLOW_DEPRECATED_DECLARATIONS_END
 
-        callOnMainThread([weakThis = WTFMove(weakThis)] {
-            if (weakThis)
-                weakThis->completeFetchData();
+            callOnMainThread([weakThis = WTFMove(weakThis)] {
+                if (weakThis)
+                    weakThis->completeFetchData();
+            });
```

Comment:
Is this `callOnMainThread` necessary, aren't we are already on the main thread?

Reply:
It is preexisting but you might be right this is not strictly needed.
I'll check whether we can remove it in a follow-up.